all-sdk/external-client-configuration.md

```diff
@@ -0,0 +1,340 @@
+  * `TEMPORAL_CONFIG_FILE` - the path to the config file. The default is `~/.config/temporalio/temporal.json`.
+* ❓ How do to grpc-meta as env var? Like for `My-Header: My-Value`
+  * Env var names are case-insensitive and don't use dashes, so we can't put header name in env var name. And we may not
+    want to canonicalize the header names even though most gRPC clients do it for you.
+  * There's no obvious env var approach to array of strings, and it can't be completely comma-delimited because header
+    values often contain commas. Would rather not force users to have an index in the env var name. Should we just not
+    support gRPC meta from env var until we think this through?
+
+## Implementation
+
+### CLI
+
+* CLI will accept `--profile`.
+  * This deprecates `--env`/`--env-file`
+    * 💭 Why deprecate?
+      * The `env` name is confusing with environment variables. Note we do not use the term "environment" anywhere in
+        this document except for environment variables.
+      * The `env` setting uses YAML files in `~/.config/temporalio/temporal.yaml`.
+    * Error if both options are present, or in the case of default, if there are default files for both.
+      * 💭 Why error in the case of both default env and profile being present instead of merge?
+        * This is the new way and we should not try to support both at the same time, it can get confusing.
+  * The default is `default`, and this can also be set via `TEMPORAL_PROFILE` env var.
+* CLI will accept `--config-file`.
+  * Default is `~/.config/temporalio/temporal.json`, also can be set via `TEMPORAL_CONFIG_FILE` env var.
+* CLI will have a whole new set of `profile` commands that operate similar to `env` commands today.
+  * Will not go into detail in this proposal, but it's very similar.
+  * We will strictly validate keys when setting whereas we did not with `env`.
+* Unlike `env`/`--env`, this does not blindly apply any field as any CLI argument.
+  * 💭 Why?
+    * We want strict validation of config values. The idea that you can set a `workflow-id` as part of this profile does
+      not make sense. These are client options only, declarative options for other things can be discussed separately.
+* CLI should defer to the Go SDK for loading/using the config/profile.
+
+### SDKs
+
+SDKs must support loading client configuration from optional profile name and optional config name. This needs to
+include a way to make sure that user-defaults can be set.
+
+Possible language-specific ideas are below. Note, they are not exact and the naming is all subject to change, they are
+simply ideas for discussion. Many of the approaches below are based on several rewritings after seeing how they don't
+work in some SDKs. The overall approach is to just make client and connection options loadable from configuration
+without overthinking shortcuts or merging or hierarchies.
```

Comment:
Not mentioned here but the actual loading and parsing of the file in Core SDKs should be done by Core I think, so that we can centralize:

* Logic of default profile / file location
* Merging of env vars
* Read/write the files

Since it's already a proto core can just return that after loading / merging.

Reply:
Makes sense, but I don't think there's a writing step to centralize, and I still expect users to be able to use the protos themselves if they want. And I think people want to manipulate the results, so I think we should centralize all of those things in Core, but not the conversion of the proto result to the client options, that should remain in lang.

Reply:
Yeah agree